### PR TITLE
Add option to generate source rpm package

### LINKED ISSUE #321: src rpm not copied out of docker
The script copies out the binary but not the source

```
+ exit 0
Processing files: wazuh-agent-3.10.2-100.x86_64
Checking for unpackaged file(s): /usr/lib/rpm/check-files /build_wazuh/rpmbuild/BUILDROOT/wazuh-agent-3.10.2-100.x86_64
Wrote: /build_wazuh/rpmbuild/SRPMS/wazuh-agent-3.10.2-100.src.rpm
Wrote: /build_wazuh/rpmbuild/RPMS/x86_64/wazuh-agent-3.10.2-100.x86_64.rpm
Executing(%clean): /bin/sh -e /var/tmp/rpm-tmp.CxNReN
+ umask 022
+ cd /build_wazuh/rpmbuild/BUILD
+ cd wazuh-agent-3.10.2
+ rm -fr /build_wazuh/rpmbuild/BUILDROOT/wazuh-agent-3.10.2-100.x86_64
+ exit 0
+ rpm_file=wazuh-agent-3.10.2-100.x86_64.rpm
+ pkg_path=/build_wazuh/rpmbuild/RPMS/x86_64
+ [[ no == \y\e\s ]]
+ mv /build_wazuh/rpmbuild/RPMS/x86_64/wazuh-agent-3.10.2-100.x86_64.rpm /var/local/wazuh
Package wazuh-agent-3.10.2-100.x86_64.rpm added to /tmp.
```
```
[root@server-01 rpms]#  ll /tmp/*.rpm
-rw-r--r--. 1 root root 9562224 Oct  1 17:34 /tmp/wazuh-agent-3.10.2-100.x86_64.rpm
```

### PR DESCRIPTION
**Pull request by DFolchA**
Hello team,

This PR closes #321, I have added an option to generate the `src.rpm` package to the RPM generation script.

Tests were done for:

- [x] Agent generation
- [x] Manager generation
- [x] Api generation

Regards, 
Daniel Folch

### CODE REVIEW

File: rpms/README.md [hunk 1]
```diff
@@ -24,30 +24,34 @@ To build an RPM package, you need to download this repository and use the `gener
...
+        -t,   --target <target>     [Required] Target package to build [manager/api/agent].
+        -a,   --architecture <arch> [Optional] Target architecture of the package [x86_64/i386].
+        -r,   --revision <rev>      [Optional] Package revision that append to version e.g. x.x.x-rev
+        -l,   --legacy              [Optional] Build package for CentOS 5.
+        -s,   --store <path>        [Optional] Set the destination path of package. By default, an output folder will be created.
+        -j,   --jobs <number>       [Optional] Number of parallel jobs when compiling.
+        -p,   --path <path>         [Optional] Installation path for the package. By default: /var.
+        -d,   --debug               [Optional] Build the binaries with debug symbols and create debuginfo packages. By default: no.
+        -c,   --checksum <path>     [Optional] Generate checksum on the desired path (by default, if no path is specified it will be generated on the same directory than the package).
+        -src, --source              [Optional] Generate the source package in the destination directory
```
Comment (BraulioV) [inline rpms/README.md]: This option should be `--src` only.

File: rpms/README.md [hunk 2]
```diff
@@ -24,30 +24,34 @@ To build an RPM package, you need to download this repository and use the `gener
...
     ```
-    * To build a wazuh-manager package for x86_64, revision 3821 and store it in `/tmp`:
-        `# ./generate_rpm_package.sh -b v3.8.2 -s /tmp -t manager -a x86_64 -r 3821`.
-    * To build a wazuh-agent package for i386 with `-j15`, revision 3 and store it in `/tmp`:
-        `# ./generate_rpm_package.sh -b v3.8.2 -s /tmp -t agent -a x86_64 -j 15 -r 3`.
-    * To build a wazuh-api package from branch 3.9 and store it in `/tmp`:
-        `# ./generate_rpm_package.sh -b 3.9 -s /tmp -t api -a x86_64 -r 0`.
-    * To build a wazuh-manager x86_64 package for `/opt/ossec` directory and store it in `/tmp`:
-        `# ./generate_rpm_package.sh -b v3.8.2 -s /tmp -t manager -a x86_64 -r 0.1 -p /opt`.
+    * To build a wazuh-manager package for `x86_64`, revision `myrev` and store it in `/tmp`:
```
Comment (BraulioV) [inline rpms/README.md]: Change the revision to a number based revision. For example: `2.el6`.

File: rpms/build.sh [hunk 3]
```diff
@@ -51,11 +52,15 @@ $linux rpmbuild --define "_topdir ${rpm_build_dir}" --define "_threads ${threads
...
 
-rpm_file="wazuh-${build_target}-${wazuh_version}-${package_release}.${architecture_target}.rpm"
-pkg_path="${rpm_build_dir}/RPMS/${architecture_target}"
+rpm_file="wazuh-${build_target}-${wazuh_version}-${package_release}*"
+pkg_path="${rpm_build_dir}/RPMS"
 
 if [[ "${checksum}" == "yes" ]]; then
-    cd ${pkg_path} && sha512sum ${rpm_file} > /var/local/checksum/${rpm_file}.sha512
+    find ${pkg_path} -maxdepth 2 -type f -name "${rpm_file}" -exec sh -c 'sha512sum {} > {}.sha512 && mv {}.sha512 /var/local/checksum'  \;
 fi
```
Comment (BraulioV) [inline rpms/build.sh]: This will include the path where the package is stored in the .sha512 file. Revert the changes.

File: rpms/build.sh [hunk 4]
```diff
@@ -51,11 +52,15 @@ $linux rpmbuild --define "_topdir ${rpm_build_dir}" --define "_threads ${threads
         --define "_debugenabled ${debug}" --target ${architecture_target} \
         -ba ${rpm_build_dir}/SPECS/${package_name}.spec
 
-rpm_file="wazuh-${build_target}-${wazuh_version}-${package_release}.${architecture_target}.rpm"
-pkg_path="${rpm_build_dir}/RPMS/${architecture_target}"
+rpm_file="wazuh-${build_target}-${wazuh_version}-${package_release}*"
```
Comment (BraulioV) [inline rpms/build.sh]: I prefer to add a new variable like `src_rpm_file` with the path to that package and then, use it to compute the checksum and extract it without using `find`.

File: rpms/build.sh [hunk 5]
```diff
@@ -51,11 +52,15 @@ $linux rpmbuild --define "_topdir ${rpm_build_dir}" --define "_threads ${threads
...
+pkg_path="${rpm_build_dir}/RPMS"
 
 if [[ "${checksum}" == "yes" ]]; then
-    cd ${pkg_path} && sha512sum ${rpm_file} > /var/local/checksum/${rpm_file}.sha512
+    find ${pkg_path} -maxdepth 2 -type f -name "${rpm_file}" -exec sh -c 'sha512sum {} > {}.sha512 && mv {}.sha512 /var/local/checksum'  \;
 fi
 
-mv ${pkg_path}/${rpm_file} /var/local/wazuh
\ No newline at end of file
+if [ "${src}" == "yes" ]; then
```
Comment (BraulioV) [inline rpms/build.sh]: Use a single `=`

File: rpms/generate_rpm_package.sh [hunk 6]
```diff
@@ -141,17 +141,18 @@ help() {
...
+    echo "    -t,   --target <target>     [Required] Target package to build [manager/api/agent]."
+    echo "    -a,   --architecture <arch> [Optional] Target architecture of the package [x86_64/i386]."
+    echo "    -r,   --revision <rev>      [Optional] Package revision that append to version e.g. x.x.x-rev"
+    echo "    -l,   --legacy              [Optional] Build package for CentOS 5."
+    echo "    -s,   --store <path>        [Optional] Set the destination path of package. By default, an output folder will be created."
+    echo "    -j,   --jobs <number>       [Optional] Number of parallel jobs when compiling."
+    echo "    -p,   --path <path>         [Optional] Installation path for the package. By default: /var."
+    echo "    -d,   --debug               [Optional] Build the binaries with debug symbols and create debuginfo packages. By default: no."
+    echo "    -c,   --checksum <path>     [Optional] Generate checksum on the desired path (by default, if no path is specified it will be generated on the same directory than the package)."
+    echo "    -src, --source              [Optional] Generate the source package in the destination directory"
```
Comment (BraulioV) [inline rpms/generate_rpm_package.sh]: This option should be `--src` only.

File: rpms/generate_rpm_package.sh [hunk 7]
```diff
@@ -241,6 +242,10 @@ main() {
                 help 1
             fi
             ;;
+        "-src"|"--source")
```
Comment (BraulioV) [inline rpms/generate_rpm_package.sh]: This option should be `--src` only.

File: rpms/README.md (hunk 1 above)
Comment (DFolchA) [reply rpms/README.md]: Done in https://github.com/wazuh/wazuh-packages/pull/322/commits/1c0fc432f8a6a01bc925799249c0761244cb4f1f

File: rpms/README.md (hunk 2 above)
Comment (DFolchA) [reply rpms/README.md]: Done in https://github.com/wazuh/wazuh-packages/pull/322/commits/1c0fc432f8a6a01bc925799249c0761244cb4f1f

File: rpms/build.sh (hunk 3 above)
Comment (DFolchA) [reply rpms/build.sh]: Done in https://github.com/wazuh/wazuh-packages/pull/322/commits/1c0fc432f8a6a01bc925799249c0761244cb4f1f

File: rpms/build.sh (hunk 4 above)
Comment (DFolchA) [reply rpms/build.sh]: Done in https://github.com/wazuh/wazuh-packages/pull/322/commits/1c0fc432f8a6a01bc925799249c0761244cb4f1f

File: rpms/build.sh (hunk 5 above)
Comment (DFolchA) [reply rpms/build.sh]: Done in https://github.com/wazuh/wazuh-packages/pull/322/commits/1c0fc432f8a6a01bc925799249c0761244cb4f1f

File: rpms/generate_rpm_package.sh (hunk 6 above)
Comment (DFolchA) [reply rpms/generate_rpm_package.sh]: Done in https://github.com/wazuh/wazuh-packages/pull/322/commits/1c0fc432f8a6a01bc925799249c0761244cb4f1f

File: rpms/generate_rpm_package.sh (hunk 7 above)
Comment (DFolchA) [reply rpms/generate_rpm_package.sh]: Done in https://github.com/wazuh/wazuh-packages/pull/322/commits/1c0fc432f8a6a01bc925799249c0761244cb4f1f

File: rpms/README.md (hunk 1 above)
Comment (BraulioV) [reply rpms/README.md]: I think is better `--src` than `-src`.

File: rpms/generate_rpm_package.sh (hunk 6 above)
Comment (BraulioV) [reply rpms/generate_rpm_package.sh]: I think is better `--src` than `-src`

File: rpms/build.sh (hunk 5 above)
Comment (BraulioV) [reply rpms/build.sh]: Check the syntax of the script. Where you are using `[[ expr ]]` you must use `==` when comparing variables, but if you use `[ expr ]` you must use `=`. In the commit 1c0fc43 you changed this. 

Can you review it again and changed only those lines where there's something like `[ x == y ]`?

File: rpms/generate_rpm_package.sh [hunk 8]
```diff
@@ -141,17 +141,18 @@ help() {
...
+    echo "    -t,   --target <target>     [Required] Target package to build [manager/api/agent]."
+    echo "    -a,   --architecture <arch> [Optional] Target architecture of the package [x86_64/i386]."
+    echo "    -r,   --revision <rev>      [Optional] Package revision that append to version e.g. x.x.x-rev"
+    echo "    -l,   --legacy              [Optional] Build package for CentOS 5."
+    echo "    -s,   --store <path>        [Optional] Set the destination path of package. By default, an output folder will be created."
+    echo "    -j,   --jobs <number>       [Optional] Number of parallel jobs when compiling."
+    echo "    -p,   --path <path>         [Optional] Installation path for the package. By default: /var."
+    echo "    -d,   --debug               [Optional] Build the binaries with debug symbols and create debuginfo packages. By default: no."
+    echo "    -c,   --checksum <path>     [Optional] Generate checksum on the desired path (by default, if no path is specified it will be generated on the same directory than the package)."
+    echo "    -src                        [Optional] Generate the source package in the destination directory"
```
Comment (BraulioV) [inline rpms/generate_rpm_package.sh]: This line is outdated.

File: rpms/generate_rpm_package.sh [hunk 9]
```diff
@@ -61,12 +61,12 @@ build_rpm() {
...
         if [[ "${TARGET}" == "manager" ]] && [[ "${LEGACY}" == "yes" ]]; then
-            MAJOR_MINOR="$(echo ${VERSION} | cut -d "v" -f 2)"
+            MAJOR_MINOR="$(echo ${VERSION}|cut -d . -f 1-2)"
             echo "major minor is $MAJOR_MINOR"
-            if [[ "${MAJOR_MINOR}" > "3.9" ]] || [[ "${MAJOR_MINOR}" == "3.9" ]]; then
+            if [[ $(echo "$MAJOR_MINOR 3.9" | awk '{print ($1 > $2)}') ]] || [[ "${MAJOR_MINOR}" == "3.9" ]]; then
                 echo "Wazuh Manager is not supported for CentOS 5 from v3.9.0."
                 echo "Version to build: ${VERSION}."
-                exit 1
+                return 1
```
Comment (BraulioV) [inline rpms/generate_rpm_package.sh]: This return should be a `clean 1`.

File: rpms/generate_rpm_package.sh [hunk 10]
```diff
@@ -61,12 +61,12 @@ build_rpm() {
         VERSION="$(cat ${SOURCES_DIRECTORY}/src/VERSION | cut -d 'v' -f 2)"
         echo "Version is $VERSION"
         if [[ "${TARGET}" == "manager" ]] && [[ "${LEGACY}" == "yes" ]]; then
-            MAJOR_MINOR="$(echo ${VERSION} | cut -d "v" -f 2)"
+            MAJOR_MINOR="$(echo ${VERSION}|cut -d . -f 1-2)"
```
Comment (BraulioV) [inline rpms/generate_rpm_package.sh]: This won't ever work. If `VERSION` is `v3.10.0`, for example, line 64 generates this: `MAJOR_MINOR = v3.10` and the `if` from 66 won't work. 

To extract the major minor, you must run this: `echo ${VERSION} | cut -d v -f 2 | cut -d . -f 1-2`

Comment (DFolchA) [reply rpms/generate_rpm_package.sh]: The `v` from the version is removed at line 61 so it should not be necessary to remove it again.

File: rpms/generate_rpm_package.sh (hunk 8 above)
Comment (DFolchA) [reply rpms/generate_rpm_package.sh]: Done in https://github.com/wazuh/wazuh-packages/pull/322/commits/7ce7042530bc906ff9d6dc27c4296d77e592ad22

File: rpms/generate_rpm_package.sh (hunk 9 above)
Comment (DFolchA) [reply rpms/generate_rpm_package.sh]: In this commit https://github.com/wazuh/wazuh-packages/pull/322/commits/7ce7042530bc906ff9d6dc27c4296d77e592ad22, I changed the `exit 1` and `clean 1` calls in build and build_rpm functions for `return 1`  this way when something fails, these functions will return 1 and this line will call to clean:
https://github.com/wazuh/wazuh-packages/blob/7ce7042530bc906ff9d6dc27c4296d77e592ad22/rpms/generate_rpm_package.sh#L264

File: rpms/generate_rpm_package.sh [hunk 11]
```diff
@@ -7,7 +7,7 @@
 # and/or modify it under the terms of the GNU General Public
 # License (version 2) as published by the FSF - Free Software
 # Foundation.
-
+set -x
```
Comment (BraulioV) [inline rpms/generate_rpm_package.sh]: This line is unnecessary and should be removed.

File: rpms/build.sh [hunk 12]
```diff
@@ -51,11 +58,13 @@ $linux rpmbuild --define "_topdir ${rpm_build_dir}" --define "_threads ${threads
...
+    cd ${src_path} && sha512sum ${src_file} > /var/local/checksum/${src_file}.sha512
+fi
+
+if [[ "${src}" == "yes" ]]; then
+    extract_path="${rpm_build_dir}"
 fi
 
-mv ${pkg_path}/${rpm_file} /var/local/wazuh
\ No newline at end of file
+find ${extract_path} -maxdepth 3 -type f -name "${file_name}*" -exec mv {} /var/local/wazuh \;
```
Comment (BraulioV) [inline rpms/build.sh]: This line extracts the src.rpm package even if the parameter is set or not. Check this out:
```shellsession
$ ls
build.sh  CentOS  generate_rpm_package.sh  README.md  SPECS
$ ./generate_rpm_package.sh -b v3.10.2 -t agent -j 8 -s $(pwd)
...
$ ls
build.sh  CentOS  generate_rpm_package.sh  README.md  SPECS  wazuh-agent-3.10.2-1.src.rpm
```
In addition, this doesn't extract the binary rpm package, just the sources.

Finally, I would remove this line completely and use something like this:
```
mv ${pkg_path}/${rpm_file} /var/local/wazuh
mv ${src_path}/${src_file} /var/local/wazuh
```

File: rpms/build.sh [hunk 13]
```diff
@@ -51,11 +58,13 @@ $linux rpmbuild --define "_topdir ${rpm_build_dir}" --define "_threads ${threads
         --define "_debugenabled ${debug}" --target ${architecture_target} \
         -ba ${rpm_build_dir}/SPECS/${package_name}.spec
 
-rpm_file="wazuh-${build_target}-${wazuh_version}-${package_release}.${architecture_target}.rpm"
-pkg_path="${rpm_build_dir}/RPMS/${architecture_target}"
-
 if [[ "${checksum}" == "yes" ]]; then
     cd ${pkg_path} && sha512sum ${rpm_file} > /var/local/checksum/${rpm_file}.sha512
+    cd ${src_path} && sha512sum ${src_file} > /var/local/checksum/${src_file}.sha512
```
Comment (BraulioV) [inline rpms/build.sh]: This line will compute the checksum of the src package always. The script should only compute the checksum of that file only if the `--src` parameter is set.

File: rpms/generate_rpm_package.sh (hunk 11 above)
Comment (DFolchA) [reply rpms/generate_rpm_package.sh]: Fixed in d20150a.

File: rpms/build.sh (hunk 12 above)
Comment (DFolchA) [reply rpms/build.sh]: Fixed in d20150a.

File: rpms/build.sh (hunk 13 above)
Comment (DFolchA) [reply rpms/build.sh]: Fixed in d20150a.
